Add hooks to customize the order receipt generation

Konamiman opened the pull request:
### Changes proposed in this Pull Request:

Adds a few extensibility points to [the receipt rendering engine]( https://github.com/woocommerce/woocommerce/pull/43502 )to allow customizing the generated receipts. More precisely:

* A `woocommerce_printable_order_receipt_data` filter to customize the data that gets combined with the receipt template in order to produce the actual receipt.
* A `woocommerce_printable_order_receipt_line_item_display_data` filter to customize the rendering of each order line item. `wp_kses_post` is applied to the HTML returned by this filter for security.
* A `woocommerce_printable_order_receipt_css` filter to customize the CSS styles embedded in the generated receipt.

See the documentation blocks for each hook in [the ReceiptRenderingEngine class](https://github.com/woocommerce/woocommerce/blob/add_hooks_for_receipt_rendering/plugins/woocommerce/src/Internal/ReceiptRendering/ReceiptRenderingEngine.php) for details.

Also in order to support these changes:

- Receipt data is now accessed through a `$data` array in the template (data is not extracted to variables anymore).
- CSS styles are moved to a separate template file.
- All the relevant HTML elements in the receipt get an explicit `id` attribute (helpful when customizing the CSS).

Closes #46122.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

1. Create an order with at least one product.

2. Create a receipt for the order via code or via REST API (see the testing instructions of [the receipts rendering engine pull request]( https://github.com/woocommerce/woocommerce/pull/43502 ) for details), TLDR using `curl`: 

```
curl -X POST -u <user>:<password> http://localhost/wp-json/wc/v3/orders/<order id>/receipt?force_new=true
```

3. Open the URL for the generated receipt in your browser, verify that it renders as expected.

4. Now add the following code snippet at the end of `woocommerce.php` or using [the Code Snippets plugin](https://wordpress.org/plugins/code-snippets/):

```php
add_filter( 'woocommerce_printable_order_receipt_data', function( $data, $order ) {
	$data['texts']['amount_paid_section_title'] = 'How much??';
	$data['notes'][] = 'Order id: ' . $order->get_id() . ' - note added via filter.';
	$data['payment_info'] = [
		'payment_method' => 'woocommerce_payments',
		'card_icon'    => base64_encode( file_get_contents( WC_ABSPATH . "src/Internal/ReceiptRendering/CardIcons/diners.svg" ) ),
		'card_last4'   => '1234',
		'app_name'     => 'some_app_name',
		'aid'          => 'some_file_name',
		'account_type' => 'some_account_type'
	];
	return $data;
}, 10, 2 );

add_filter( 'woocommerce_printable_order_receipt_line_item_display_data', function( $line_item_display_data, $line_item_data, $order ) {
    if( $line_item_data['type'] === 'product' ) {
        $product_id = $line_item_data['item']->get_product()->get_id();
        $order_id = $order->get_id();
        $line_item_display_data['inner_html'] =
            //To test the wp_kses_post, try this instead:
	        //"<script>alert('YOU\'VE BEEN H4CK3D');</script>";
            "<td>${line_item_display_data['row_index']}: ($product_id in $order_id) " . substr( $line_item_display_data['inner_html'], 4 );
    }
    else {
        $line_item_display_data['tr_attributes']['class'] = 'non_product';
    }
    return $line_item_display_data;
}, 10, 3 );

add_filter( 'woocommerce_printable_order_receipt_css', function( $css, $order ) {
    $amount = $order->get_total();
    if($amount < 100) {
        $color = '00F000';
    }
    else if($amount < 1000) {
        $color = 'F0F000';
    }
    else {
        $color = 'F00000';
    }

    return $css . "table#line_items { background-color:#$color; } tr.non_product {font-style: italic}";
}, 10, 2 );
```

5. Regenerate the order receipt (don't forget to add the `force_true` argument to the request).

6. Open the new receipt and verify that it has changed as follows:

* The "Amount paid" section title is now "How much??"
* The background color of the line items table is now green if the total amount of the order is less than 100, yellow if the amount is between 100 and 1000, and green if it's more than 1000.
* Line items corresponding to products have now a prefix that includes the row number, the product id and the order id.
* Line items that don't correspond to products are rendered in italics.
* There's one fake order note.
* There's fake payment information.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

Konamiman committed:
Add hooks to customize the order receipt generation.

Also:

- Move CSS to a separate template
- Access receipt data through a $data array in the template
  (don't extract the variables anymore)

Konamiman committed:
Add changelog file

Konamiman committed:
Add missing linter disabling in the new CSS template

helgatheviking commented on plugins/woocommerce/src/Internal/ReceiptRendering/Templates/order-receipt.php:
```diff
@@ -3,99 +3,63 @@
-			<?php
-		}
+	foreach ( $data['formatted_line_items'] as $formatted_line_item ) {
+		echo "<tr>$formatted_line_item</tr>";
```
@Konamiman Can this `<tr>` element have a filter for a custom class? That's what Product Bundles, Composites, and Mix and Match all use to help style the resulting rows.

Konamiman replied:
You can use the `woocommerce_printable_order_receipt_css` filter to add custom CSS for these elements using the `table#line_items tr` selector, would that help?

helgatheviking replied:
I need to add rules to certain line item TRs but not all. For example a bundle line item will have different styling than the bundled line items. Traditionally we have a different class of the row for targeting. How could we do that without?

Konamiman replied:
Ok, makes sense. I've modified the `woocommerce_printable_order_receipt_formatted_line_item` filter so that now it includes the `<tr>` element too (so now the filter must return the entire table row).

helgatheviking replied:
Is that preferable to adding a separate filter for just the class? This seem really hard if say both myself and another plugin need to look at the class as now we need to use regex/string matching to see if the `<tr>` already has a `class=` attribute.

Konamiman replied:
I have replaced the `woocommerce_printable_order_receipt_formatted_line_item` filter with a `woocommerce_printable_order_receipt_line_item_display_data` one that contains both the attributes for the `<tr>` element and the HTML that gets rendered inside it. I hope it's comprehensive enough now.

Konamiman committed:
Modify line items filter to include the tr element too

Konamiman committed:
Replace woocommerce_printable_order_receipt_formatted_line_item filter
with the more complete woocommerce_printable_order_receipt_line_item_display_data

helgatheviking commented on plugins/woocommerce/src/Internal/ReceiptRendering/ReceiptRenderingEngine.php:
```diff
@@ -149,15 +161,19 @@ public function generate_receipt( $order, $expiration_date = null, bool $force_n
-			$formatted_line_items[] = $formatted_line_item;
+			$line_item_display_data = apply_filters( 'woocommerce_printable_order_receipt_line_item_display_data', $line_item_display_data, $line_item_data, $order );
+			$attributes             = '';
+			foreach ( $line_item_display_data['tr_attributes'] as $attribute_name => $attribute_value ) {
```
@Konamiman Am I correct in thinking that to add a row class you'd filter the `woocommerce_printable_order_receipt_line_item_display_data` value and pass back `$data['tr_attributes']['class'] = 'my-class';` what would be the best way to ensure that Plugin A and Plugin B can both add a class without stomping on each other?

helgatheviking replied:
And is there a need for escaping/sanitizing the attributes before outputting them?

Konamiman replied:
You can check if there's already a `'class'` key in `$data['tr_attributes']`, and if so you add a space and your class name to the existing value. In one line:

```php
$data['tr_attributes']['class'] = trim(($data['tr_attributes']['class'] ?? '') . ' my-class')
```

> And is there a need for escaping/sanitizing the attributes before outputting them?

That's a good point. I've added an `esc_attr` for the attribute values received from the hook.

Konamiman committed:
Add esc_attr for attribute values coming from hook

jorgeatorres commented on plugins/woocommerce/src/Internal/ReceiptRendering/ReceiptRenderingEngine.php:
```diff
@@ -197,20 +276,24 @@ private function get_order_data( WC_Order $order ): array {
 					( wc_get_product( $line_item_product->get_parent_id() )->get_name() ) . '. ' . $line_item_product->get_attribute_summary() :
 					$line_item_product->get_name();
```
This wasn't touched by this PR, but be aware that `$line_item_product` can be `false`. For example, if the product has been removed since the order was created. There would still be a name to be used in the line item itself, but the product won't exist and this would produce a fatal error.

I believe we should check that ` $line_item_product` is not false before using it.

Konamiman replied:
So if `$line_item_product` is false, fallback to simply `$line_item->get_name()`?

jorgeatorres replied:
Yes. Sounds good!

Konamiman replied:
I've moved this fix here: https://github.com/woocommerce/woocommerce/pull/47300

jorgeatorres commented on plugins/woocommerce/src/Internal/ReceiptRendering/ReceiptRenderingEngine.php:
```diff
@@ -115,11 +115,90 @@ public function generate_receipt( $order, $expiration_date = null, bool $force_n
+		 * @param WC_Order $order The order for which the receipt is being generated.
+		 * @returns array The updated set of data.
+		 *
+		 * @since 8.9.0
```
This (and other occurrences) would have to be updated now to `9.0.0` 😄

jorgeatorres reviewed (changes requested):
Hi @Konamiman!

Overall this looks good to me (left some minimal feedback) but I wonder if we're being too open/lax with the filters and the customizations here.

For instance, by using `woocommerce_printable_order_receipt_line_item_display_data` you could even include scripts or tags such as `onclick` or `onload` in HTML attributes, which feels too risky to me. Could we pass things through `wp_kses()` before the final output is created and maybe limit to a reasonable set of tags?
I understand that 3rd party code has a lot of freedom and could break things in a different way, but we should still try to prevent some broken attribute name from rendering the receipt rendering useless.

I also wonder if there's any need to customize things to this level by using filters. I guess most people will only want to change the name of certain line items or the CSS classes applied to those items, but not necessarily add arbitrary HTML.
I might be wrong, though, but then wouldn't it make more sense to allow the template to be overridden entirely? Not suggesting we do that here, just something to think about.

Let me know what your thoughts are!

Konamiman commented:
> I wonder if we're being too open/lax with the filters and the customizations here

I think it's fine, the purpose of hooks in general is to provide as much customization as possible and I'm pretty sure we have such "risky" (as in: very flexible) filters anywhere else. Also these receipts are intended for being sent to printing servers anyway; I don't think it would make much sense trying to inject JS or something there.

> Could we pass things through wp_kses() before the final output is created and maybe limit to a reasonable set of tags?

So something like `<td>` and `<p>`? And with which attributes?

> wouldn't it make more sense to allow the template to be overridden entirely?

Doesn't this kind of contradict your first statement? Overriding the entire template is the maximum degree of being open/lax 🙂

Anyway, using a custom template is something that can be implemented in the future if needed as a complement to these filters.

jorgeatorres commented:
> Also these receipts are intended for being sent to printing servers anyway; I don't think it would make much sense trying to inject JS or something there.

I get that, though "printing service" might be a regular browser and Cmd + P. I don't think this is a blocker, as we do have other risky filters in the codebase, but we've been moving towards sanitizing/escaping even after filters are applied precisely to prevent weird things from happening. At least trying to prevent some XSS wouldn't hurt and we do have the sanitizing functions to do that.

> So something like <td> and <p>? And with which attributes?

I guess I'm mostly worried about <script> tags in the unlikely scenario the pages are actually rendered in a browser. I wouldn't want a script being able to read private info from the receipt. Other HTML is probably fine.

> Doesn't this kind of contradict your first statement?

Yes. Such is life. I was just thinking that we're still providing pretty much that but via various filters, so we might as well support overriding the template, but that's not something we have to consider right now.

Konamiman commented:
> I guess I'm mostly worried about <script> tags in the unlikely scenario the pages are actually rendered in a browser. I wouldn't want a script being able to read private info from the receipt. Other HTML is probably fine.

How could that be implemented? Is there some kind of "negative wp_kses" function in the codebase?

Konamiman committed:
Verify if $line_item->get_product() is false (will happen if
the product no longer exists) and use line item's get_name() if so
to render the line item text.

Konamiman committed:
Appli wp_kses_post to the HTML returned by the
woocommerce_printable_order_receipt_line_item_display_data filter.

Konamiman committed:
Update @since tags

jorgeatorres commented on plugins/woocommerce/src/Internal/ReceiptRendering/ReceiptRenderingEngine.php:
```diff
@@ -115,11 +115,91 @@ public function generate_receipt( $order, $expiration_date = null, bool $force_n
+		$formatted_line_items = array();
+		$row_index            = 0;
+		foreach ( $data['line_items'] as $line_item_data ) {
+			$quantity_data          = isset( $line_item_data['quantity'] ) ? " × ${line_item_data['quantity']}" : '';
+			$line_item_display_data = array(
+				'inner_html'    => "<td>${line_item_data['title']}$quantity_data</td><td>${line_item_data['amount']}</td>",
+				'tr_attributes' => array(),
+				'row_index'     => $row_index++,
+			);
```
Use of `${var}` instead of `{$var}` is now deprecated, so we better update these now.

jorgeatorres commented on plugins/woocommerce/src/Internal/ReceiptRendering/ReceiptRenderingEngine.php:
```diff
@@ -115,11 +115,91 @@ public function generate_receipt( $order, $expiration_date = null, bool $force_n
+				$attribute_value = esc_attr( $attribute_value );
+				$attributes     .= " $attribute_name=\"$attribute_value\"";
+			}
+			$formatted_line_items[] = wp_kses_post( "<tr$attributes>${line_item_display_data['inner_html']}</tr>" );
```
Here too.

jorgeatorres reviewed (changes requested):
@Konamiman! Thanks for making the updates. This looks good to me. Left some minor feedback for something I didn't notice during the first review, but that should be very straightforward to fix. Thanks again!

Konamiman committed:
Replace ${ with {$ in interpolated strings to avoid deprecation warnings

jorgeatorres reviewed (approved):
LGTM. Thank you @Konamiman!

By the way, I think you might need to go through another round of "fun with CI" due to failing checks before we can merge.

Konamiman committed:
Merge branch 'trunk' into add_hooks_for_receipt_rendering

Konamiman committed:
Revert "Verify if $line_item->get_product() is false (will happen if"

This reverts commit dbf772cf9e1765b852f21092c6bda4d790046d5c.

Konamiman committed:
Merge branch 'trunk' into add_hooks_for_receipt_rendering
